File: ios/MKWidgetCard/ContactSaver.swift
import Contacts
import Foundation
import React

@objc(ContactSaver)
class ContactSaver: NSObject {
  @objc
  static func requiresMainQueueSetup() -> Bool {
    false
  }

  @objc(saveContact:resolver:rejecter:)
  func saveContact(
    _ payload: NSDictionary,
    resolver resolve: @escaping RCTPromiseResolveBlock,
    rejecter reject: @escaping RCTPromiseRejectBlock
  ) {
    let store = CNContactStore()

    let save = {
      let contact = CNMutableContact()
      let fullName = Self.stringValue(payload["fullName"])
      let nameParts = Self.splitName(fullName)

      contact.givenName = nameParts.givenName
      contact.familyName = nameParts.familyName
      contact.organizationName = Self.stringValue(payload["company"])
      contact.jobTitle = Self.stringValue(payload["title"])
      contact.departmentName = Self.stringValue(payload["department"])

      let phone = Self.stringValue(payload["phone"])
      if !phone.isEmpty {
        contact.phoneNumbers = [
          CNLabeledValue(label: CNLabelPhoneNumberMobile, value: CNPhoneNumber(stringValue: phone))
        ]
      }

      let email = Self.stringValue(payload["email"])
      if !email.isEmpty {
        contact.emailAddresses = [
          CNLabeledValue(label: CNLabelWork, value: email as NSString)
        ]
      }

      let address = Self.stringValue(payload["address"])
      if !address.isEmpty {
        let postalAddress = CNMutablePostalAddress()
        postalAddress.street = address
        contact.postalAddresses = [
          CNLabeledValue(label: CNLabelWork, value: postalAddress)
        ]
      }

      let website = Self.stringValue(payload["website"])
      let linkedin = Self.stringValue(payload["linkedin"])
      let facebook = Self.stringValue(payload["facebook"])
      let zaloUrl = Self.stringValue(payload["zaloUrl"])
      let whatsappUrl = Self.stringValue(payload["whatsappUrl"])
      let telegramUrl = Self.stringValue(payload["telegramUrl"])
      let bio = Self.stringValue(payload["bio"])
      contact.urlAddresses = [
        website.isEmpty ? nil : CNLabeledValue(label: CNLabelURLAddressHomePage, value: website as NSString),
        linkedin.isEmpty ? nil : CNLabeledValue(label: "LinkedIn", value: linkedin as NSString),
        facebook.isEmpty ? nil : CNLabeledValue(label: "Facebook", value: facebook as NSString),
        zaloUrl.isEmpty ? nil : CNLabeledValue(label: "Zalo", value: zaloUrl as NSString),
        whatsappUrl.isEmpty ? nil : CNLabeledValue(label: "WhatsApp", value: whatsappUrl as NSString),
        telegramUrl.isEmpty ? nil : CNLabeledValue(label: "Telegram", value: telegramUrl as NSString),
        bio.isEmpty ? nil : CNLabeledValue(label: "Bio", value: bio as NSString)
      ].compactMap { $0 }

      let zalo = Self.stringValue(payload["zalo"])
      let whatsapp = Self.stringValue(payload["whatsapp"])
      let telegram = Self.stringValue(payload["telegram"])
      contact.instantMessageAddresses = [
        zalo.isEmpty ? nil : CNLabeledValue(label: "Zalo", value: CNInstantMessageAddress(username: zalo, service: "Zalo")),
        whatsapp.isEmpty ? nil : CNLabeledValue(label: "WhatsApp", value: CNInstantMessageAddress(username: whatsapp, service: "WhatsApp")),
        telegram.isEmpty ? nil : CNLabeledValue(label: "Telegram", value: CNInstantMessageAddress(username: telegram, service: "Telegram"))
      ].compactMap { $0 }

      contact.socialProfiles = [
        linkedin.isEmpty ? nil : CNLabeledValue(label: "LinkedIn", value: CNSocialProfile(urlString: linkedin, username: "", userIdentifier: "", service: "LinkedIn")),
        facebook.isEmpty ? nil : CNLabeledValue(label: "Facebook", value: CNSocialProfile(urlString: facebook, username: "", userIdentifier: "", service: "Facebook"))
      ].compactMap { $0 }

      if let imageData = Self.photoData(from: Self.stringValue(payload["avatarUrl"])) {
        contact.imageData = imageData
      }

      let request = CNSaveRequest()
      request.add(contact, toContainerWithIdentifier: nil)

      do {
        try store.execute(request)
        resolve(["saved": true])
      } catch {
        reject("contacts_save_failed", "Unable to save contact.", error)
      }
    }

    switch CNContactStore.authorizationStatus(for: .contacts) {
    case .authorized, .limited:
      save()
    case .notDetermined:
      store.requestAccess(for: .contacts) { granted, error in
        if granted {
          save()
        } else {
          reject("contacts_permission_denied", "Contacts permission was denied.", error)
        }
      }
    case .denied, .restricted:
      reject("contacts_permission_denied", "Contacts permission was denied.", nil)
    @unknown default:
      reject("contacts_permission_denied", "Contacts permission is unavailable.", nil)
    }
  }

  private static func stringValue(_ value: Any?) -> String {
    guard let value = value as? String else {
      return ""
    }

    return value.trimmingCharacters(in: .whitespacesAndNewlines)
  }

  private static func photoData(from value: String) -> Data? {
    guard !value.isEmpty else {
      return nil
    }

    if value.lowercased().hasPrefix("data:image/"),
       let commaIndex = value.firstIndex(of: ",") {
      let base64 = String(value[value.index(after: commaIndex)...])
      return Data(base64Encoded: base64)
    }

    guard let url = URL(string: value) else {
      return nil
    }

    if url.isFileURL {
      return try? Data(contentsOf: url)
    }

    guard ["http", "https"].contains(url.scheme?.lowercased() ?? "") else {
      return nil
    }

    return try? Data(contentsOf: url)
  }

  private static func splitName(_ fullName: String) -> (givenName: String, familyName: String) {
    let parts = fullName
      .split(whereSeparator: { $0.isWhitespace })
      .map(String.init)

    guard parts.count > 1 else {
      return (fullName, "")
    }

    return (parts.dropLast().joined(separator: " "), parts.last ?? "")
  }
}

File: android/app/src/main/java/com/mktech/widgetcard/ContactSaver.kt
package com.mktech.widgetcard

import android.Manifest
import android.content.ContentProviderOperation
import android.content.pm.PackageManager
import android.net.Uri
import android.provider.ContactsContract
import android.util.Base64
import com.facebook.react.bridge.Promise
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReactContextBaseJavaModule
import com.facebook.react.bridge.ReactMethod
import com.facebook.react.bridge.ReadableMap
import java.net.HttpURLConnection
import java.net.URL

class ContactSaver(private val reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    override fun getName(): String = "ContactSaver"

    @ReactMethod
    fun saveContact(payload: ReadableMap, promise: Promise) {
        if (reactContext.checkSelfPermission(Manifest.permission.WRITE_CONTACTS) != PackageManager.PERMISSION_GRANTED) {
            promise.reject("contacts_permission_denied", "Contacts permission was denied.")
            return
        }

        try {
            val fullName = payload.cleanString("fullName")
            if (fullName.isBlank()) {
                promise.reject("contacts_save_failed", "Full name is required.")
                return
            }

            val (givenName, familyName) = splitName(fullName)
            val ops = ArrayList<ContentProviderOperation>()

            ops.add(
                ContentProviderOperation.newInsert(ContactsContract.RawContacts.CONTENT_URI)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_TYPE, null)
                    .withValue(ContactsContract.RawContacts.ACCOUNT_NAME, null)
                    .build()
            )

            ops.add(
                ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                    .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                    .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.StructuredName.CONTENT_ITEM_TYPE)
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.DISPLAY_NAME, fullName)
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.GIVEN_NAME, givenName)
                    .withValue(ContactsContract.CommonDataKinds.StructuredName.FAMILY_NAME, familyName)
                    .build()
            )

            addOrganization(ops, payload)
            addPhone(ops, payload.cleanString("phone"))
            addEmail(ops, payload.cleanString("email"))
            addAddress(ops, payload.cleanString("address"))
            addWebsite(ops, payload.cleanString("website"), ContactsContract.CommonDataKinds.Website.TYPE_WORK)
            addWebsite(ops, payload.cleanString("linkedin"), ContactsContract.CommonDataKinds.Website.TYPE_OTHER, "LinkedIn")
            addWebsite(ops, payload.cleanString("facebook"), ContactsContract.CommonDataKinds.Website.TYPE_OTHER, "Facebook")
            addWebsite(ops, payload.cleanString("zaloUrl"), ContactsContract.CommonDataKinds.Website.TYPE_OTHER, "Zalo")
            addWebsite(ops, payload.cleanString("whatsappUrl"), ContactsContract.CommonDataKinds.Website.TYPE_OTHER, "WhatsApp")
            addWebsite(ops, payload.cleanString("telegramUrl"), ContactsContract.CommonDataKinds.Website.TYPE_OTHER, "Telegram")
            addInstantMessage(ops, "Zalo", payload.cleanString("zalo"))
            addInstantMessage(ops, "WhatsApp", payload.cleanString("whatsapp"))
            addInstantMessage(ops, "Telegram", payload.cleanString("telegram"))
            addNote(ops, payload.cleanString("note").ifBlank { payload.cleanString("bio") })
            addPhoto(ops, payload.cleanString("avatarUrl"))

            val results = reactContext.contentResolver.applyBatch(ContactsContract.AUTHORITY, ops)
            promise.resolve(mapOf("saved" to true, "rawContactId" to (results.firstOrNull()?.uri?.lastPathSegment ?: "")))
        } catch (error: Exception) {
            promise.reject("contacts_save_failed", "Unable to save contact.", error)
        }
    }

    private fun addOrganization(ops: ArrayList<ContentProviderOperation>, payload: ReadableMap) {
        val company = payload.cleanString("company")
        val title = payload.cleanString("title")
        val department = payload.cleanString("department")
        if (company.isBlank() && title.isBlank() && department.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Organization.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Organization.COMPANY, company)
                .withValue(ContactsContract.CommonDataKinds.Organization.TITLE, title)
                .withValue(ContactsContract.CommonDataKinds.Organization.DEPARTMENT, department)
                .build()
        )
    }

    private fun addPhone(ops: ArrayList<ContentProviderOperation>, phone: String) {
        if (phone.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Phone.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Phone.NUMBER, phone)
                .withValue(ContactsContract.CommonDataKinds.Phone.TYPE, ContactsContract.CommonDataKinds.Phone.TYPE_MOBILE)
                .build()
        )
    }

    private fun addEmail(ops: ArrayList<ContentProviderOperation>, email: String) {
        if (email.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Email.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Email.ADDRESS, email)
                .withValue(ContactsContract.CommonDataKinds.Email.TYPE, ContactsContract.CommonDataKinds.Email.TYPE_WORK)
                .build()
        )
    }

    private fun addAddress(ops: ArrayList<ContentProviderOperation>, address: String) {
        if (address.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.StructuredPostal.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.StructuredPostal.FORMATTED_ADDRESS, address)
                .withValue(ContactsContract.CommonDataKinds.StructuredPostal.TYPE, ContactsContract.CommonDataKinds.StructuredPostal.TYPE_WORK)
                .build()
        )
    }

    private fun addWebsite(
        ops: ArrayList<ContentProviderOperation>,
        url: String,
        type: Int,
        label: String? = null
    ) {
        if (url.isBlank()) return

        val builder = ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
            .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
            .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Website.CONTENT_ITEM_TYPE)
            .withValue(ContactsContract.CommonDataKinds.Website.URL, url)
            .withValue(ContactsContract.CommonDataKinds.Website.TYPE, type)

        if (!label.isNullOrBlank()) {
            builder.withValue(ContactsContract.CommonDataKinds.Website.LABEL, label)
        }

        ops.add(builder.build())
    }

    private fun addInstantMessage(ops: ArrayList<ContentProviderOperation>, service: String, username: String) {
        if (username.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Im.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Im.DATA, username)
                .withValue(ContactsContract.CommonDataKinds.Im.TYPE, ContactsContract.CommonDataKinds.Im.TYPE_OTHER)
                .withValue(ContactsContract.CommonDataKinds.Im.PROTOCOL, ContactsContract.CommonDataKinds.Im.PROTOCOL_CUSTOM)
                .withValue(ContactsContract.CommonDataKinds.Im.CUSTOM_PROTOCOL, service)
                .build()
        )
    }

    private fun addNote(ops: ArrayList<ContentProviderOperation>, note: String) {
        if (note.isBlank()) return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Note.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Note.NOTE, note)
                .build()
        )
    }

    private fun addPhoto(ops: ArrayList<ContentProviderOperation>, avatarUrl: String) {
        val bytes = loadPhotoBytes(avatarUrl) ?: return

        ops.add(
            ContentProviderOperation.newInsert(ContactsContract.Data.CONTENT_URI)
                .withValueBackReference(ContactsContract.Data.RAW_CONTACT_ID, 0)
                .withValue(ContactsContract.Data.MIMETYPE, ContactsContract.CommonDataKinds.Photo.CONTENT_ITEM_TYPE)
                .withValue(ContactsContract.CommonDataKinds.Photo.PHOTO, bytes)
                .build()
        )
    }

    private fun loadPhotoBytes(value: String): ByteArray? {
        if (value.isBlank()) return null

        return when {
            value.startsWith("data:image/", ignoreCase = true) -> {
                val commaIndex = value.indexOf(',')
                if (commaIndex < 0) null else Base64.decode(value.substring(commaIndex + 1), Base64.DEFAULT)
            }
            value.startsWith("http://", ignoreCase = true) || value.startsWith("https://", ignoreCase = true) -> {
                val connection = (URL(value).openConnection() as HttpURLConnection).apply {
                    connectTimeout = 8000
                    readTimeout = 8000
                    instanceFollowRedirects = true
                    setRequestProperty("User-Agent", "MKWidgetCard/1.0")
                }

                try {
                    if (connection.responseCode !in 200..299) {
                        null
                    } else {
                        connection.inputStream.use { it.readBytes() }
                    }
                } finally {
                    connection.disconnect()
                }
            }
            value.startsWith("file://", ignoreCase = true) || value.startsWith("content://", ignoreCase = true) -> {
                reactContext.contentResolver.openInputStream(Uri.parse(value))?.use { it.readBytes() }
            }
            else -> null
        }
    }

    private fun ReadableMap.cleanString(key: String): String {
        return if (hasKey(key) && !isNull(key)) {
            getString(key)?.trim().orEmpty()
        } else {
            ""
        }
    }

    private fun splitName(fullName: String): Pair<String, String> {
        val parts = fullName.split(Regex("\\s+")).filter { it.isNotBlank() }
        if (parts.size <= 1) return fullName to ""

        return parts.dropLast(1).joinToString(" ") to parts.last()
    }
}
